Add containers dashboard node utilization donut chart

zeari opened the pull request:
![donut](https://cloud.githubusercontent.com/assets/11256940/11629019/377757a4-9cfe-11e5-8346-5e68bea19a05.png)


~~Depends on #5481 and #5338~~(Both merged)

@abonas @simon3z

~~(Ill get a better screenshot when Ill accumulate more metrics)~~

zeari commented on app/services/container_dashboard_service.rb:
```diff
@@ -78,4 +81,82 @@ def providers
+                              sum(derived_vm_numvcpus) as total_cores,
+                              sum(derived_memory_available) as total_mem,
+                              cast(timestamp as date) as day")
+    metrics = metrics.order("day").group("day")
```
@Fryguy Please have a look at this query with ```VimPerformanceDaily```

Fryguy replied:
If there is no `@ems`, then won't line 123 end up bringing back *all* EMSes including cloud and infra?

Fryguy replied:
Also, the `cast(timestamp as date) as day` and `metrics.order("day").group("day")` ignores the time zone from the time profile.  The timestamps are UTC so you would have to cast into that that timezone for that timeprofile in the query before chunking by day.

Fryguy replied:
I really think this should just be done in Ruby.  I dislike dropping down to SQL for everything, and as is happening, as soon as it gets slightly more complicated you have to drop down to Ruby anyway.  I believe the chunking by day is already done in Ruby code, probably in VimPerformanceDaily, but I'd have to check.

zeari commented:
@miq-bot add_label providers/containers
@miq-bot add_label ui

abonas commented on app/assets/javascripts/application.js:
```diff
@@ -40,8 +40,13 @@
 //= require controllers/ops/log_collection_form_controller
 //= require controllers/ops/tenant_form_controller
 //= require controllers/ops/tenant_quota_form_controller
+//= require controllers/container_dashboard/util/util-module
```
what is util-module?
the name is a bit vague, and it's also under util folder, and under it there's also dashboard-utils...

zeari replied:
> what is util-module?
> the name is a bit vague, and it's also under util folder, and under it there's also dashboard-utils...

@abonas Yeah, I guess its a little misleading. util-module is just a one line file defining the module.
 containers_dashboard needs a few resuable assets. dashboard-utils and ```dashboardUtilsFactory``` within it should probably be something like ```dashboardObjFactory```

abonas commented on app/assets/javascripts/controllers/container_dashboard/card/heatmaps/heatmaps-card-directive.js:
```diff
@@ -0,0 +1,23 @@
+            }
+        ]
+    };
+}]);
```
heatmap card is not part of latest pf that's shipped with miq?

zeari replied:
> heatmap card is not part of latest pf that's shipped with miq?

```pf-heatmap``` and ```pf-heatmap-legend``` are. Check ```/static/heatmaps-card.html```.
(@jeff-phillips-18 correct me if I'm wrong)

jeff-phillips-18 replied:
Correct @zeari , the heatmap and the heatmap legend are part of A-PF, but the card is not since it's easy enough to put charts in cards. However, there is some debate on making specific cards for different charts in future A-PF releases.

abonas commented on app/assets/javascripts/controllers/container_dashboard/container_dashboard_controller.js:
```diff
@@ -54,10 +95,21 @@ angular.module('containerDashboard', ['ui.bootstrap', 'patternfly', 'patternfly.
+
+                    // Heatmaps
+                    $scope.nodeCpuUsage.data = data.heatmaps.nodeCpuUsage;
+                    $scope.nodeCpuUsage.loadingDone = true;
```
is this ever set to false? why is this variable needed?

zeari replied:
> is this ever set to false? why is this variable needed? (and the rest of the similar vars)

Missed that but it still works because ```ng-if``` is ok with undefined variables. Will be fixed.

abonas commented on app/models/container_replicator.rb:
```diff
@@ -8,8 +8,18 @@ class ContainerReplicator < ActiveRecord::Base
+  has_many :metrics,                :as => :resource
+  has_many :metric_rollups,         :as => :resource
+  has_many :vim_performance_states, :as => :resource
+  delegate :my_zone,                :to => :ext_management_system
```
Why is this change for metrics needed for container replicator entity?
And if it's needed, why is it part of the dashboard PR that adds node utilization?

zeari replied:
@abonas Its changes from #5338 that I needed here.
However #5587 was merged so I no longer need to be rebased on #5338.
Ill push a properly based version.

abonas replied:
> @abonas Its changes from #5338 that I needed here.
> However #5587 was merged so I no longer need to be rebased on #5338.
> Ill push a properly based version.

yep, see https://github.com/ManageIQ/manageiq/pull/5338#issuecomment-160901889

abonas commented on spec/javascripts/controllers/container_dashboard/container_dashboard_controller_spec.js:
```diff
@@ -61,7 +61,33 @@ describe('containerDashboardController', function() {
+                    used:12,
+                    xData:[
+                        "dates",
+                        "2015-11-26"
```
what are those hardcoded dates?

zeari replied:
> what are those hardcoded dates?

@abonas Mock data for specs...

abonas replied:
> @abonas Mock data for specs...

yap, I missed the "spec" folder location...

himdel replied:
Now that JSONFixture is working, I guess we should use it here as well?

zeari replied:
> Now that JSONFixture is working, I guess we should use it here as well?

#5781 has to be merged first :)

himdel replied:
Well, if you copy the same changes here, it should merge cleanly regardless of the order. But sure, up to you :).

abonas commented:
@serenamarie125 , Please look at the screenshot here.
There are 2 sections named "Node Utilization".
I think we should refine it to make it more clear to the users.
For example - "Aggregated node utilization" for the left one, and something also more focused for the right one.
Thoughts?

abonas commented on app/models/container_service.rb:
```diff
@@ -12,6 +12,21 @@ class ContainerService < ActiveRecord::Base
+  has_many :metrics,                :as => :resource
+  has_many :metric_rollups,         :as => :resource
+  has_many :vim_performance_states, :as => :resource
+  delegate :my_zone,                :to => :ext_management_system
```
same question as https://github.com/ManageIQ/manageiq/pull/5648#discussion_r46251735

abonas commented on app/services/container_dashboard_service.rb:
```diff
@@ -1,13 +1,16 @@
 class ContainerDashboardService
-  def initialize(provider_id)
+  def initialize(provider_id, controller)
```
if the controller is passed only to extract the user, why not pass just the user? (I didn't notice other usages of the controller in the service)

zeari replied:
@abonas gonna need it anyway in #5142

serenamarie125 commented:
Sounds good @abonas, please change left card to "Aggregated Node Utilization"

zeari commented:
> Sounds good @abonas, please change left card to "Aggregated Node Utilization"

Fixed. New screenshot above.

zeari commented on app/services/container_dashboard_service.rb:
```diff
@@ -86,4 +89,65 @@ def providers
+        :yData => ["used"] + used_mem.values.map {|m| (m / 1024.0).round(1)} # convert to GB
+      }
+    }
+  end
```
@Fryguy Second attempt, please review.

zeari commented on app/services/container_dashboard_service.rb:
```diff
@@ -86,4 +89,65 @@ def providers
+      date = metric.timestamp.strftime("%Y-%m-%d")
+      used_cpu[date] += (metric.derived_vm_numvcpus * metric.cpu_usage_rate_average / 100.0)
+      used_mem[date] += (metric.derived_memory_used)
+      total_cpu[date] += (metric.derived_vm_numvcpus)
```
@simon3z just wondering, why does ```derived_vm_numvcpus``` produce fractions like ```2.3454``` and not round numbers?  Doesn't this field represent the number of cores or cpus on a host?

simon3z replied:
@zeari it shouldn't produce fractions. I've never seen that behavior (both in `metrics` and `metric_rollups`).

Make sure you don't have old entries and use latest master (may be related to #5587)

simon3z commented on app/services/container_dashboard_service.rb:
```diff
@@ -86,4 +89,69 @@ def providers
+
+    metrics.each do |metric|
+      date = metric.timestamp.strftime("%Y-%m-%d")
+      used_cpu[date] += (metric.derived_vm_numvcpus * metric.cpu_usage_rate_average / 100.0)
```
@zeari you should use `v_derived_cpu_total_cores_used`

zeari replied:
> v_derived_cpu_total_cores_used

Thats much better. Where can I see which fields in ```MetricRollup``` fill up for provider\node rollups?

zeari commented on app/services/container_dashboard_service.rb:
```diff
@@ -86,4 +89,69 @@ def providers
+    }
+  end
+
+  def node_utilization
```
@kbrock Can you tell me if method is ok metric-wise?

zeari replied:
@Fryguy Your previous comments were addressed here

kbrock replied:
@gtanzillo @jrafanie do those metrics match what you are expecting?

gtanzillo replied:
@kbrock These metrics are what I would expect to be used based on what's being shown on the dashboard charts.

jrafanie replied:
thanks @gtanzillo, I didn't know the answer :question:

zeari commented on app/services/container_dashboard_service.rb:
```diff
@@ -86,4 +89,70 @@ def providers
+    else
+      {}
+    end
+  end
```
I made a small changes to round the stats and to appease Rubocop. Everything else is the same.

zeari commented:
#5481 was merged and Id like for this to get merged as well.
@himdel Please review

fyi @abonas @simon3z I believe all of your comments were addressed

himdel commented on app/assets/javascripts/controllers/container_dashboard/util/chart-config-util.js:
```diff
@@ -0,0 +1,22 @@
+var chartConfig = {
```
Uh, can we refactor this please? We really don't want to introduce new globals.. maybe use `angular.module('containerDashboard').constant('chartConfig', { ... })` (and inject when needed)?

The other option would be to make it explicit by putting it under the ManageIQ global object, but I'm guessing `.constant` (or `.value`) works, right?

zeari replied:
> The other option would be to make it explicit by putting it under the ManageIQ global object, but I'm guessing .constant (or .value) works, right?

Now its a part of ```charts-mixin.js```

simon3z commented on app/services/container_dashboard_service.rb:
```diff
@@ -160,4 +161,43 @@ def heatmaps
+  def node_utilization
+    resource_ids = @ems.present? ? [@ems.id] : ManageIQ::Providers::ContainerManager.all.pluck(:id)
+    metrics = VimPerformanceDaily.find_entries({:tz => @controller.current_user.get_timezone})
+    metrics = metrics.where(:resource_type => 'ExtManagementSystem', :resource_id => resource_ids)
```
Unbounded? Don't you want to limit this to a certain timestamp? (last 30 days, etc.) Isn't this going to dump all the data since when you added the provider?

zeari replied:
> Unbounded? Don't you want to limit this to a certain timestamp? (last 30 days, etc.) Isn't this going to dump all the data since when you added the provider?

Right. I lost it somewhere along the way. Good catch

zeari replied:
> Unbounded? Don't you want to limit this to a certain timestamp? (last 30 days, etc.) Isn't this going to dump all the data since when you added the provider?

Comments addressed. Repushed.

simon3z commented on app/services/container_dashboard_service.rb:
```diff
@@ -160,4 +161,43 @@ def heatmaps
+    used_cpu = Hash.new { |hash, key| hash[key] = 0 }
+    used_mem = used_cpu.dup
+    total_cpu = used_cpu.dup
+    total_mem = used_cpu.dup
```
This doesn't seem readable (easily understandable) at first sight. If you want to spare the repetition of `Hash.new { |hash, key| hash[key] = 0 }` just make it a function.

zeari replied:
> This doesn't seem readable (easily understandable) at first sight. If you want to spare the repetition of Hash.new { |hash, key| hash[key] = 0 } just make it a function.

@simon3z how about:
```Ruby
    used_cpu = Hash.new(0)
    used_mem = Hash.new(0)
    total_cpu = Hash.new(0)
    total_mem = Hash.new(0)
```

It should also set the default value to zero but looks much nicer.

simon3z commented on app/services/container_dashboard_service.rb:
```diff
@@ -160,4 +161,44 @@ def heatmaps
+    resource_ids = @ems.present? ? [@ems.id] : ManageIQ::Providers::ContainerManager.all.pluck(:id)
+    metrics = VimPerformanceDaily.find_entries({:tz => @controller.current_user.get_timezone})
+    metrics = metrics.where(:resource_type => 'ExtManagementSystem', :resource_id => resource_ids)
+    metrics = metrics.where("timestamp > ?", 30.days.ago)
```
@zeari @alongoldboim any added value in keeping all the `where` clause in the same call? E.g.:

    metrics.where(:resource_type => 'ExtManagementSystem', :resource_id => resource_ids,
                  "timestamp > ?", 30.days.ago))

zeari replied:
> @zeari @alongoldboim any added value in keeping all the where clause in the same call? E.g.:

When I spilt the line I assumed that since this query executes only at line 177 and that I can form it anyway I like before that

alongoldboim replied:
@simon3z In general rails doesn't do the query and get the object until it needs to, in that case splitting the 
"metrics.where(:resource_type => 'ExtManagementSystem', :resource_id => resource_ids)" and "metrics = metrics.where("timestamp > ?", 30.days.ago)" wont make any difference because rails doesn't preform the query between the two calls (though i don't understand what the reason for splitting it).

zeari replied:
> (though i don't understand what the reason for splitting it).

What would be the syntax for keeping it in one line?

This causes an exception:
``` 
metrics.where(:resource_type => 'ExtManagementSystem', :resource_id => resource_ids, "timestamp > ?", 30.days.ago)
```

zeari replied:
> any added value in keeping all the where clause in the same call? E.g.:

@simon3z @alongoldboim It could be:
```
metrics = metrics.where('resource_type = :type and resource_id in (:resource_ids) and timestamp > :min_time',
                            :type => 'ExtManagementSystem', :resource_ids => resource_ids, :min_time => 30.days.ago)
```

which doesn't look as nice to me.

alongoldboim replied:
@zeari I think it looks better and its more readable, Would advise to use this format for longer queries instead of splitting them up.

simon3z commented:
@zeari we need tests for `node_utilization` (mostly to make sure that the filtering is correct).

zeari commented on spec/services/container_dashboard_service_spec.rb:
```diff
@@ -0,0 +1,103 @@
+      })
+    end
+  end
+end
```
@simon3z Added tests, have a look here please.

zeari commented:
@himdel I changed this patch, did some refactoring and added proper tests. Please re-review.

@simon3z Added metric tests on the ruby end and implemented @alongoldboim's comment.

himdel commented on app/assets/javascripts/controllers/container_dashboard/card/heatmaps/heatmaps-card-directive.js:
```diff
@@ -1,4 +1,4 @@
-angular.module('miq.card').directive('heatmapsCard', ['ChartsDataMixin' ,function(chartsDataMixin) {
+angular.module('miq.card').directive('heatmapsCard', ['chartsMixin' ,function(chartsDataMixin) {
```
Maybe rename `chartsDataMixin` to `chartsMixin` as well?

zeari replied:
> Maybe rename chartsDataMixin to chartsMixin as well?

Right. Missed that.

himdel commented on spec/services/container_dashboard_service_spec.rb:
```diff
@@ -1,4 +1,5 @@
 require "spec_helper"
+require 'recursive-open-struct'
```
We should probably add `recursive-open-struct` to the Gemfile. Right now it works only because kubeclient and openshift-client depend on it.

zeari replied:
> We should probably add recursive-open-struct to the Gemfile. Right now it works only because kubeclient and openshift-client depend on it.

I wasnt aware of this and I can see that it ```require 'recursive-open-struct'``` also appears in another file. Ill open a separate PR for this.

himdel replied:
Great, thanks! :)

zeari commented on spec/services/container_dashboard_service_spec.rb:
```diff
@@ -25,4 +26,74 @@
+        }
+      )
+    end
+  end
```
@abonas @simon3z A review of this test is the last thing I need to get this merged. Please review.

simon3z commented on app/services/container_dashboard_service.rb:
```diff
@@ -140,4 +139,44 @@ def heatmaps
     }
   end
+
+  def node_utilization
```
Minor: @zeari it seems that this is more `ems_utilization`, or `ems_aggregated_utilization`. In the method you are never dealing with nodes.

simon3z commented on spec/services/container_dashboard_service_spec.rb:
```diff
@@ -25,4 +26,74 @@
+      ems_openshift.metric_rollups << current_metric
+      ems_openshift.metric_rollups << old_metric
+      ems_kubernetes.metric_rollups << current_metric.dup
+      ems_kubernetes.metric_rollups << old_metric.dup
```
@zeari I think that to be on the safe-side you should use different values for the two providers (e.g. 4 cpus vs 2, 4096 avail mem vs 2048, etc.). Because at the moment a bug that aggregates twice just one of the two providers would pass the `node_utilization_all_providers` test.

simon3z commented:
There are two minor outstanding comments. But overall LGTM :+1: cc @chessbyte @blomquisg @abonas

zeari commented:
Fixed comments by @simon3z. Please merge.

simon3z commented:
LGTM :+1: cc @chessbyte ready for merge

chessbyte commented:
@martinpovolny @dclarizio @h-kataria This is all UI code.  Please review and merge if it looks good to you.
